ops/pebble.py

```diff
@@ -1966,3 +1952,162 @@ def send_signal(self, sig: typing.Union[int, str], services: typing.List[str]):
+        for file in self._files.values():
+            os.unlink(file.name)
+
+    def feed(self, data: bytes):
+        """Provide more data to the parser to process."""
+        self._buffer += data
+        # self._run_parser_fsm()
+
+        header_terminator = b'\r\n\r\n'
+
+        while True:
+            # Try to read headers, if present.
+            if self._headers is None:
+                if self._buffer.startswith(self._boundary) and header_terminator in self._buffer:
+                    # Data is everything after the boundary, up to and including the double CRLF
+                    # of the headers.
+                    data_start = len(self._boundary)
+                    data_end = self._buffer.index(header_terminator) + len(header_terminator)
+                    data = self._buffer[data_start:data_end]
+
+                    self._headers = self._parse_headers(data)
+                    self._detect_invalid_content_disposition()
+                    self._part_type = self._prepare_for_part_body()
+
+                    # Advance the buffer to point at the data.
+                    self._buffer = self._buffer[data_end:]
+                elif self._buffer.startswith(self._terminal_boundary):
+                    # We've reached the terminal boundary; ignore any remaining data
+                    self.terminated = True
+                    break
+                else:
+                    # Insufficient data to continue
+                    break
+
+            if self._part_type == 'response':
+                next_boundary_index = self._get_next_boundary_index()
+                if next_boundary_index is not None:
+                    data = self._buffer[:next_boundary_index]
+                    self._response = json.loads(data.decode())
+                    # Advance the buffer to point at the next boundary
+                    self._buffer = self._buffer[next_boundary_index:]
+                    self._headers = None
+                else:
+                    # Insufficient data to continue
+                    break
+            elif self._part_type == 'files':
+                next_boundary_index = self._get_next_boundary_index()
+                if next_boundary_index is not None:
+                    # Next boundary's location is known; this is the final write.
+                    data = self._buffer[:next_boundary_index]
+                    with self._get_open_tempfile() as outfile:
+                        outfile.write(data)
+                    # Advance the buffer to point at the next boundary
+                    self._buffer = self._buffer[next_boundary_index:]
+                    self._headers = None
+                else:
+                    # Data present but boundary not yet found.
+                    # Write data from buffer, but don't accidentally write data which may belong
+                    # to a boundary line.
+                    safe_bound = len(self._buffer) - len(self._terminal_boundary)
```

Comment:
Nice -- this cross-boundary edge case would be easy to miss. :-) Do we have an explicit regression test for the boundary crossing the chunk size? I think we should if not.

Speaking of which, I think we should have a couple of tests of (relatively) large files. Though they don't have to be actually big, we could just set the chunk size to small, like 200 bytes, and have a 1000 byte file. The main things I'd want to test for are: boundary spanning a chunk, data more than 1 chunk, data more than a few chunks.

Reply:
It would also be good to add a few tests to `TestRealPebble` and test them against the real Go-based pebble server. Or *at least* do thorough manual testing against a real pebble via the test CLI -- but it might be easier to just add `TestRealPebble` tests anyway.

Reply:
I added a round trip push+pull real pebble test and also one test with a tiny chunk size causing data and boundaries to span multiple chunks.  Caught a small bug actually!